Interest rates and spread matrix command - economy menu

montezdesousa opened the pull request:
# Description

This PR creates a new command to analyze cross-country interest rate curves and changes.

By default it shows a chart with the 10Y G7 countries spread matrix.

Snip from comment below for better context:

```
The spread matrix provides a nice way to see the yield difference between countries. This is useful for both institutional traders and retail.

Speculators usually don't do directional trades so they're more interested in buying one bond vs selling other - essentially trading the spread between them. Central banks and economists themselves monitor these differences particularly on eurozone. For instance, Italy vs Germany is a proxy to monitor political risk in the eurozone.

The changes matrix is also a very good way to check if market open is risk on or off in the interest rates market, which relates to stock markets as well. Usually if periphery spreads in Europe are tightening against Germany market is greenish and the other way around.
```

The user will be able to:
* Select from a list of groups like G7, PIIGS, Eurozone, Middle East, Africa, Americas...
* Specify which countries to include in the matrix
* Specify the maturity to analyze
* See matrix of current spreads or 1 day changes
* Use color flag to choose matrix colors from: rgb, binary or openbb
* Use raw flag produces a colorized rich table

TODO:
- [x] Change some titles
- [x] Add more groups
- [x] Provide --color flag to choose between binary and gradient colors
- [x] Create unit tests
- [x] Add integration tests for `spread`, `ycrv`, `events` and fix wrong test cases on `test_economy.openbb` (placed warning on `ycrv` when date is passed without `--source FRED`)

`spread --group G7`
`screenshot`
![image](https://user-images.githubusercontent.com/79287829/195433385-2435b418-3e31-4da4-aa63-89eb9c8421b4.png)

`spread --raw`
![image](https://user-images.githubusercontent.com/79287829/195306372-78cfe203-a2ad-47f5-b35e-ff3c77ca1bd1.png)

`spread --group EZ --color binary`
![image](https://user-images.githubusercontent.com/79287829/195305678-e2505972-3371-4167-854e-febd878c3f4e.png)

`spread --group PIIGS --color binary`
![image](https://user-images.githubusercontent.com/79287829/195305840-589439e9-a91e-448a-99ec-3a17c2db7665.png)

`spread --countries portugal,spain,italy,united states --maturity 5Y --change=True`
![image](https://user-images.githubusercontent.com/79287829/195306154-c200d41c-d483-4861-b365-bc58281947c7.png)

`spread --color openbb`
![image](https://user-images.githubusercontent.com/79287829/195437961-c535f8fc-9972-4a10-a4e4-8360b826b887.png)


- [x] Summary of the change / bug fix.
- [x] Link # issue, if applicable.
- [x] Screenshot of the feature or the bug before/after fix, if applicable.
- [x] Relevant motivation and context. 
- [x] List any dependencies that are required for this change.


# How has this been tested?

* Please describe the tests that you ran to verify your changes. 
* Provide instructions so we can reproduce. 
* Please also list any relevant details for your test configuration.
- [x] Make sure affected commands still run in terminal
- [x] Ensure the api still works
- [x] Check any related reports


# Checklist:

- [x] Update [our Hugo documentation](https://openbb-finance.github.io/OpenBBTerminal/) following [these guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/tree/main/website).
- [x] Update our tests following [these guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/tree/main/tests).
- [x] Make sure you are following our [CONTRIBUTING guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/blob/main/CONTRIBUTING.md).
- [x] If a feature was added make sure to add it to the corresponding [scripts file](https://github.com/OpenBB-finance/OpenBBTerminal/tree/main/scripts).


# Others
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.

jose-donato commented:
This looks awesome

JerBouma commented:
Hi @montezdesousa, can you give some financial context, how would an institution utilise spread differences and what is generally the incurred costs (making capitalising on a spread irrelevant)? Also curious about further use cases, would I be interested in these spreads if I just wish to cover interest rate risk on a portfolio?

montezdesousa commented:
> Hi @montezdesousa, can you give some financial context, how would an institution utilise spread differences and what is generally the incurred costs (making capitalising on a spread irrelevant)? Also curious about further use cases, would I be interested in these spreads if I just wish to cover interest rate risk on a portfolio?

The spread matrix provides a nice way to see the yield difference between countries. This is useful for both institucional traders and retail.

Speculators usually don't do directional trades so they're more interested in buying one bond vs selling other - essentially trading the spread between them. Central banks and economists themselves monitor these differences particularly on eurozone. For instance, Italy vs Germany is a proxy to monitor political risk in the eurozone.

The changes matrix is also a very good way to check if market open is risk on or off in the interest rates market. Usually if periphery spreads in Europe are tightening against Germany market is greenish and the other way around.

We can discuss more applications if you want.

jose-donato commented:
Would be cool to set a gradient of colors (color map) for the matrix colors - to better understand how high/low a value is compared to another (I did something similar for `crypto/ov/hm`)

JerBouma commented:
@montezdesousa That are some very interesting insights, thank you. Does the current implementation only show the **current** spreads or is it also possible to look at spreads in the past?

montezdesousa commented:
> @montezdesousa That are some very interesting insights, thank you. Does the current implementation only show the **current** spreads or is it also possible to look at spreads in the past?

Only current, but I think this could be easily extended to historic if we have a reliable historic source. The last time I tried to use investpy historic bond data it was not working.

montezdesousa commented:
> Would be cool to set a gradient of colors (color map) for the matrix colors - to better understand how high/low a value is compared to another (I did something similar for `crypto/ov/hm`)

nice tip, will try to include it. Thanks @jose-donato

JerBouma commented:
> > @montezdesousa That are some very interesting insights, thank you. Does the current implementation only show the **current** spreads or is it also possible to look at spreads in the past?
> 
> Only current, but I think this could be easily extended to historic if we have a reliable historic source. The last time I tried to use investpy historic bond data it was not working.

Good to know, well this is a really great start. The main issue I would now see is that people do not know how big a spread is truly abnormal and what is not during certain market conditions but that's still TBD.

montezdesousa commented:
> > > @montezdesousa That are some very interesting insights, thank you. Does the current implementation only show the **current** spreads or is it also possible to look at spreads in the past?
> > 
> > 
> > Only current, but I think this could be easily extended to historic if we have a reliable historic source. The last time I tried to use investpy historic bond data it was not working.
> 
> Good to know, well this is a really great start. The main issue I would now see is that people do not know how big a spread is truly abnormal and what is not during certain market conditions but that's still TBD.

Agreed, we have monthly series for 10Y bond rates in `fred` actually! https://fred.stlouisfed.org/series/IRLTLT01ITM156N

jmaslek commented:
Can we call this something like `spread` instead of `matrix`

jmaslek commented:
> > @montezdesousa That are some very interesting insights, thank you. Does the current implementation only show the **current** spreads or is it also possible to look at spreads in the past?
> 
> Only current, but I think this could be easily extended to historic if we have a reliable historic source. The last time I tried to use investpy historic bond data it was not working.

One #2559 is merged, one could create the desired historical spread using a fred series :P

montezdesousa commented:
> Can we call this something like `spread` instead of `matrix`

Sure

jmaslek commented:
Lets also make sure this ends up in scripts/test_economy.openbb :D

montezdesousa commented:
> Lets also make sure this ends up in scripts/test_economy.openbb :D

oops, there now. Also added integration tests for ycrv and events and fixed some.

JerBouma commented:
> > > @montezdesousa That are some very interesting insights, thank you. Does the current implementation only show the **current** spreads or is it also possible to look at spreads in the past?
> > 
> > 
> > Only current, but I think this could be easily extended to historic if we have a reliable historic source. The last time I tried to use investpy historic bond data it was not working.
> 
> One #2559 is merged, one could create the desired historical spread using a fred series :P

Yeah, I think we still need to think of the curation process of (among other things) FRED series. A good extension would indeed be to collect all those series **although** I saw not all are fully up to date. See: https://fred.stlouisfed.org/searchresults/?st=%20Long-Term%20Government%20Bond%20Yields

jmaslek commented on openbb_terminal/economy/investingcom_model.py:
```diff
@@ -33,6 +33,229 @@
+        "Greece",
+        "Ireland",
+        "Italy",
+        # "Latvia",
+        # "Lithuania",
+        # "Luxembourg",
```
Is there a reason certain ones are commented out instead of just not being included?

montezdesousa replied:
Yes, these countries belong to the group and Investing.com have some of them available in the website. 

There seems to be an issue with investpy library when getting the data. So I just commented them for now to avoid warning and not forget to include them when it is fixed.

montezdesousa replied:
Do u think we should keep them all so it works automatically later? Downside is the warning..

montezdesousa replied:
>  @jmaslek do u think we should keep them all so it works automatically later? Downside is the warning..

jmaslek replied:
I will leave that up to you.  Can always leave it as a TODO: to check when they are available.

jmaslek commented on openbb_terminal/economy/investingcom_model.py:
```diff
@@ -33,6 +33,229 @@
+
+@log_start_end(log=logger)
+def create_matrix(dictionary: Dict[str, Dict[str, float]]) -> pd.DataFrame:
+
```
Docstring missing

montezdesousa replied:
Added tks

jmaslek commented on openbb_terminal/economy/economy_controller.py:
```diff
@@ -1145,12 +1157,102 @@ def call_ycrv(self, other_args: List[str]):
+            dest="color",
+            type=str,
+            choices=investingcom_view.COLORS,
+            default="rgb",
```
default openbb ?

montezdesousa replied:
good for me!

jmaslek commented:
This is looking. real good :D

![Screen Shot 2022-10-13 at 11 16 38 AM](https://user-images.githubusercontent.com/18151143/195638017-b9edc715-91e5-4a5d-8b9d-a54790f13ab2.png)

When data not found, can we print the countries and not the list:

```
2022 Oct 13, 11:18 (🦋) /economy/ $ spread -m 2Y -g AMERICAS
Downloading: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7/7 [00:02<00:00,  2.65it/s]
No data for ['Mexico'].
```

montezdesousa commented:
> This is looking. real good :D
> 
> ![Screen Shot 2022-10-13 at 11 16 38 AM](https://user-images.githubusercontent.com/18151143/195638017-b9edc715-91e5-4a5d-8b9d-a54790f13ab2.png)
> 
> When data not found, can we print the countries and not the list:
> 
> ```
> 2022 Oct 13, 11:18 (🦋) /economy/ $ spread -m 2Y -g AMERICAS
> Downloading: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7/7 [00:02<00:00,  2.65it/s]
> No data for ['Mexico'].
> ```

yep, I was lazy there lol

jmaslek commented:
Last thing:  you added the website content index, but need to update main.yml so that this actually shows up

jmaslek reviewed (approved):
boooom

montezdesousa commented:
> boooom

![image](https://user-images.githubusercontent.com/79287829/195650410-46ca575c-03ba-4248-b1c6-0ada88cc6e6e.png)
